pypit/artrace.py

```diff
@@ -1359,265 +1365,543 @@ def model_tilt(slf, det, msarc, censpec=None, maskval=-999999.9,
+                wgd = np.where(ccyfit == maskval)
                 if wgd[0].size != 0:
                     continue
-                cc = np.correlate(ccyfit, yfit, mode='same')
-                params, fail = arutils.gauss_lsqfit(xfit, cc, 0.0)
-                centv = ccval + pcen - arcdet[j] - params[1]
-                if guesstilts is not None:
-                    shfit[sz-k] = guesstilts[arcdet[j], ordcen[arcdet[j], 0]-k]*float(msarc.shape[0]-1.0)
-                xtfit[sz-k] = ordcen[arcdet[j], 0] - k
-                ytfit[sz-k] = centv - shfit[sz-k]
-                etfit[sz-k] = 0.02
-                apfit[sz-k] = params[0]
-                if fail:
-                    mtfit[sz-k] = 1
-                else:
-                    pcen = int(0.5 + centv)
-                    mtfit[sz-k] = 0
-            if offchip:
-                # Don't use lines that go off the chip (could lead to a bad trace)
-                aduse[j] = False
-                badlines += 1
+                ccval = arcdet[j] + np.sum(xfit*ccyfit)/np.sum(ccyfit)
+                tstcc = False  # Once we have an array, there's no need to keep looking
+            # yfit = msarc[pcen-nspecfit:pcen+nspecfit+1,ordcen[arcdet[j],0]-k]
+            yfit = msarc[pcen-nspecfit:pcen+nspecfit+1, ordcen[arcdet[j], slitnum]-k-nsmth:ordcen[arcdet[j], slitnum]-k+nsmth+1]
+            if len(yfit.shape) == 2:
+                yfit = np.median(yfit, axis=1)
+            if np.size(yfit) == 0:
+                offchip = True
+                break
+            wgd = np.where(yfit == maskval)
+            if wgd[0].size != 0:
+                continue
+            cc = np.correlate(ccyfit, yfit, mode='same')
+            params, fail = arutils.gauss_lsqfit(xfit, cc, 0.0)
+            centv = ccval + pcen - arcdet[j] - params[1]
+            xtfit[sz-k] = ordcen[arcdet[j], slitnum] - k
+            ytfit[sz-k] = centv
+            etfit[sz-k] = 0.02
+            apfit[sz-k] = params[0]
+            if fail:
+                mtfit[sz-k] = 1
+            else:
+                pcen = int(0.5 + centv)
+                mtfit[sz-k] = 0
+        if offchip:
+            # Don't use lines that go off the chip (could lead to a bad trace)
+            aduse[j] = False
+            badlines += 1
+            trcdict = pad_dict(trcdict)
+            continue
+        for k in range(sz+1-nsmth, sz+1):
+            xtfit[sz-k] = ordcen[arcdet[j], slitnum]-k
+
+        wmask = np.where(mtfit == 0)
+        ytfit[np.where(mtfit == 1)] = maskval
+
+        # Append the trace information into the dictionary
+        trcdict["xtfit"].append(xtfit.copy())
+        trcdict["ytfit"].append(ytfit.copy())
+        trcdict["wmask"].append(wmask[0].copy())
+    trcdict["aduse"] = aduse
+    trcdict["badlines"] = badlines
+    msgs.info("Completed spectral tilt tracing".format(np.sum(aduse)))
+    return trcdict
+
+
+def trace_fweight(fimage, xinit, invvar=None, radius=3.):
```

Comment:
I was thinking of bringing this in myself.  ;)

For what it is worth, I have ported trace_crude for DESI.